### PR TITLE
[mm 699] - Reporter should receive DM when someone comments on their issue  Fixes #699

### LINKED ISSUE #699: Reporter should receive DM when someone comments on their issue
Currently the assignee of an issue receives a notification when someone comments on an issue they are assigned to. It would be beneficial if the reporter could also receive a notification.

### PR DESCRIPTION
**Pull request by sibasankarnayak**
This PR modifies appendCommentNotifications so both Assignee and Reporter will get a notification when someone comments on their issue (previously, only Assignee gets the notification)

ticket here 
Fixes #699

### CODE REVIEW

**Comment by hanzei** [comment]
@sibasankarnayak Please link the ticket this PR fixes uses the `Fixes #X` syntax as described in https://docs.google.com/document/d/1JpTkyjwsdbacV83NXgo8gRA_Zvcjk4DkCRFoiGBUVBs/edit.

File: server/webhook_worker.go
Comment (mickmister) [inline server/webhook_worker.go:13]: CI is having issues with the error `File is not 'goimports'-ed with -local`. This is happening because we are grouping external packages with our own internally defined ones here. We need to have a new line in between these imports:

```suggestion
	"github.com/andygrunwald/go-jira"

	"github.com/mattermost/mattermost-plugin-jira/server/utils"
	"github.com/mattermost/mattermost-plugin-jira/server/utils/types"
```

After this change is made, it results in three properly separated groups of imports:
- built-in Go packages first (`fmt`, `time` etc.)
- external packages (usually prefixed with `github.com`)
- packages defined in this project (contains `mattermost-plugin-jira`)

Comment (sibasankarnayak) [reply server/webhook_worker.go]: @mickmister Thankyou , this helped a lot

**Comment by sibasankarnayak** [comment]
> This PR has been automatically labelled "stale" because it hasn't had recent activity. A core team member will check in on the status of the PR to help with questions. Thank you for your contribution!
> 
> /cc @jasonblais @jfrerich @emilyacook

@levb  @mickmister  can you review this

**Comment by mickmister** [comment]
> can you review this

Yes, I will review the PR soon. Thanks for pinging us.

@sibasankarnayak FYI, by quoting the full comment above you've pinged other people that don't need to be pinged in this case. Especially when the quoted comment gets several levels deep, this can be an issue. Just be careful who you are pinging in your comments, and make sure it's only the people you want to communicate with. Thanks again for the ping to review :+1:

File: server/webhook_worker.go
Comment (mickmister) [inline server/webhook_worker.go:88]:
```suggestion
	if wh.eventTypes.ContainsAny("event_created_comment", "event_updated_comment") {
```

File: server/webhook_worker.go
Comment (mickmister) [inline server/webhook_worker.go:93]: I think we can just use `wh.JiraWebhook` here, instead of creating a new `JiraWebhook`

File: server/webhook_worker.go
Comment (mickmister) [inline server/webhook_worker.go:87]: Can we move this function to a file where we are defining methods on `*webhook`?

```suggestion
func (wh *webhook) applyReporterNotification(bb []byte, reporter *jira.User) error {
```

File: server/webhook_worker.go
Comment (mickmister) [inline server/webhook_worker.go:116]: We should make sure no this user is not already in the `wh.notifications` array. If they are in the array, we should just return instead of adding them here

**Review by mickmister — CHANGES_REQUESTED** [review]
Functionality LGTM, just one request to make sure we don't send duplicate notifications

**Comment by mickmister** [comment]
@aaronrothschild Do you think the reporter should get a notification for when the issue is updated, or only when someone comments on the issue?

File: server/webhook_worker.go
Comment (sibasankarnayak) [reply server/webhook_worker.go:116]: @mickmister 
i see the webhook will be called individually so there won't be duplicates correct me if i am wrong

**Comment by aaronrothschild** [comment]
> @aaronrothschild Do you think the reporter should get a notification for when the issue is updated, or only when someone comments on the issue?

Personally, I like getting notified when it is updated, but 0/5

File: server/webhook_worker.go
Comment (mickmister) [reply server/webhook_worker.go]: I would look through the code and find all the ways a notification can be added to this array. I believe the current state would be:

- The user was mentioned in the comment
- The user is the assignee
- The user is the reporter

So basically if they were mentioned or are the assignee, they will already be in the array here.

File: server/webhook_worker.go
Comment (mickmister) [reply server/webhook_worker.go]: According to https://community-daily.mattermost.com/core/pl/97cg76sa4bgf5raq53dat1hrwo, we actually don't want to send these notifications for comment updates. Can you please ensure we are not sending comment update notifications for mentions or assignees either?

File: server/webhook.go
Comment (mickmister) [inline server/webhook.go:212]: To keep the indentation of this function flat:

```suggestion
	if !wh.eventTypes.ContainsAny("event_created_comment") {
        return
    }
```

File: server/webhook.go
Comment (mickmister) [inline server/webhook.go:220]: Why do we need to `Unmarshal` this if we already have the `JiraWebhook` struct? Is it not populated properly?

File: server/webhook.go
Comment (mickmister) [inline server/webhook.go:222]: Is `UpdateAuthor` used when the event is comment creation, as opposed to comment update?

File: server/webhook.go
Comment (mickmister) [inline server/webhook.go:230]: Why create another variable when we already have this in `jwhook`?

Comment (mickmister) [reply server/webhook.go:230]: What's the purpose of creating `whook` here? It's not used anywhere

File: server/webhook.go
Comment (mickmister) [inline server/webhook.go:239]: Can we check the `if reporter == nil` at the top of this function instead?

File: server/webhook.go
Comment (mickmister) [inline server/webhook.go:219]: We won't ever reach this point since the initial webhook parser returns this error already in this case

File: server/webhook_worker.go
Comment (mickmister) [inline server/webhook_worker.go:66]: Can we add a newline in between each of these separate operations? Also I don't think `applyReporterNotification` needs to return an error. See my comments on that function above

```suggestion
	}

	v.applyReporterNotification(msg.Data, v.Issue.Fields.Reporter)

	if _, _, err = wh.PostNotifications(ww.p, msg.InstanceID); err != nil {
		ww.p.errorf("WebhookWorker id: %d, error posting notifications, err: %v", ww.id, err)
	}
```

Comment (mickmister) [reply server/webhook_worker.go:66]: Why do we need to pass in `msg.Data` here? All of the webhook data has already been unmarshaled into the `webhook` struct

**Review by mickmister — CHANGES_REQUESTED** [review]
Great job on this @sibasankarnayak :+1: I have a few requests

File: server/webhook_worker.go
Comment (sibasankarnayak) [reply server/webhook_worker.go]: my bad , have fixed the broken piece of code

File: server/webhook.go
Comment (sibasankarnayak) [reply server/webhook.go]: yes

File: server/webhook_worker.go
Comment (sibasankarnayak) [reply server/webhook_worker.go]: @mickmister seems a silly question.
 when you say to check if the notification array and assure the notification is already there 
 
 ```
 webhookUserNotification{
		jiraUsername:  reporter.Name,
		jiraAccountID: reporter.AccountID,
		message:       commentMessage,
		postType:      PostTypeComment,
		commentSelf:   jwhook.Comment.Self,
		recipientType: recipientTypeReporter,
	}
 ```
 this is the notification we append , so should i check the array with the username ,accountID , postType and recipientType ?

Comment (mickmister) [reply server/webhook_worker.go]: We just need to check username and accountID, since those are the identifiers for the user receiving the notification. The goal here is that one user should receive only one notification for this webhook request.

File: server/webhook_worker.go
Comment (mickmister) [inline server/webhook_worker.go:60]: It looks like we are no longer posting the notifications here

File: server/webhook.go
Comment (mickmister) [inline server/webhook.go:28]: `recipientTypeAssignee` isn't used anywhere. What is the use for introducing the `recipientType` property?

File: server/webhook_parser.go
Comment (sibasankarnayak) [inline server/webhook_parser.go:307]: @mickmister reply on this comment https://github.com/mattermost/mattermost-plugin-jira/pull/810#discussion_r760752142 

just to make it generalise the recipientType values as constant considering will be used further
if you want can revert the changes

File: server/webhook.go
Comment (sibasankarnayak) [reply server/webhook.go:28]: can refer to this comment https://github.com/mattermost/mattermost-plugin-jira/pull/810#discussion_r760757276

File: server/webhook_worker.go
Comment (sibasankarnayak) [reply server/webhook_worker.go:60]: Sorry my mistake it got removed while cleaning code

File: server/webhook_worker.go
Comment (sibasankarnayak) [reply server/webhook_worker.go]: Sure

File: server/webhook_parser.go
Comment (mickmister) [reply server/webhook_parser.go:307]: > considering will be used further

Are you saying this value is used elsewhere? Also do we need a `recipientTypeMention`?

File: server/webhook_worker.go
Comment (mickmister) [inline server/webhook_worker.go:68]: Can we add a newline after this block?

Comment (mickmister) [reply server/webhook_worker.go:68]: I'm thinking we should add a linting rule to this for the plugin repos. Do you think you can look into this Siba?

File: server/webhook_parser.go
Comment (mickmister) [reply server/webhook_parser.go:307]: What is the value used for? Is it checked somewhere in a different PR?

**Review by mickmister — CHANGES_REQUESTED** [review]
Thanks @sibasankarnayak :+1:

Something I forgot to verify here, we need a separate notification setting for this feature. There is some discussion on this here https://github.com/mattermost/mattermost-plugin-jira/pull/705#discussion_r631444153

File: server/webhook_parser.go
Comment (sibasankarnayak) [reply server/webhook_parser.go:307]: i dont see any where it is used , this was a hardcoded value for `webhookUserNotification` so created constant for that and used.

File: server/webhook_worker.go
Comment (sibasankarnayak) [reply server/webhook_worker.go:68]: @mickmister this is something new for me , can we do this in a different issue . seems may be can a cross plugin task

Comment (mickmister) [reply server/webhook_worker.go:68]: Since we're introducing this feature, we need a way for the user to toggle it. 1/5 it's within the scope of introducing the feature. If it's done as a separate issue, the other PR should be merged first and this one can use that functionality.

**Comment by mickmister** [comment]
> Something I forgot to verify here, we need a separate notification setting for this feature. There is some discussion on this here #705 (comment)

@sibasankarnayak Since we're introducing this feature, we need a way for the user to toggle it. 1/5 it's within the scope of introducing the feature to add the separate setting. If not, the other PR should be merged first and this one can use that functionality.

**Comment by sibasankarnayak** [comment]
> > Something I forgot to verify here, we need a separate notification setting for this feature. There is some discussion on this here #705 (comment)
> 
> @sibasankarnayak Since we're introducing this feature, we need a way for the user to toggle it. 1/5 it's within the scope of introducing the feature to add the separate setting. If not, the other PR should be merged first and this one can use that functionality.

@mickmister just to be clear we want a command to be added which can be on/off by user to control this feature ?

**Comment by mickmister** [comment]
> @mickmister just to be clear we want a command to be added which can be on/off by user to control this feature ?

Yes there is a similar command `/jira settings notifications`. We will split this into three commands
`/jira settings notifications assignee`
`/jira settings notifications mention`
`/jira settings notifications reporter`

Some discussion on the data model behind this at this comment https://github.com/mattermost/mattermost-plugin-jira/pull/705#discussion_r631444153

**Comment by sibasankarnayak** [comment]
> > @mickmister just to be clear we want a command to be added which can be on/off by user to control this feature ?
> 
> Yes there is a similar command `/jira settings notifications`. We will split this into three commands `/jira settings notifications assignee` `/jira settings notifications mention` `/jira settings notifications reporter`
> 
> Some discussion on the data model behind this at this comment [#705 (comment)](https://github.com/mattermost/mattermost-plugin-jira/pull/705#discussion_r631444153)

@mickmister when we refer to mention is it implies for creating the DM post for plugin

**Comment by mickmister** [comment]
@sibasankarnayak When we refer to "mention notifications", we mean the Jira user was mentioned in a comment on Jira. In this case, we should send a bot DM to that connected user in Mattermost, if they have their notification settings set to receive "mention" notifications.

**Comment by sibasankarnayak** [comment]
@hanzei @mickmister can you review the code , have added the notification for assignee ,  mention,reporter too

**Comment by sibasankarnayak** [comment]
@mickmister can you share your feedback on this .

picked up the incomplete PR from @jupriano  for issue #507 once this will be merge as notification setting for reporter and assignee is been added in this PR

File: go.mod
Comment (mickmister) [inline go.mod:21]: The changes in this file and `go.sum` seem unrelated to the purpose of this PR. Can you please revert these changes?

**Review by mickmister — CHANGES_REQUESTED** [review]
Great work @sibasankarnayak! I have some requests, the main one being that we should introduce the new notification settings as `*bool` in order to be backwards compatible with their current preferences. Please see the following comments for context:

https://github.com/mattermost/mattermost-plugin-jira/pull/705#discussion_r631444153

https://github.com/mattermost/mattermost-plugin-jira/pull/792#discussion_r693156228

File: server/command_test.go
Comment (sibasankarnayak) [reply server/command_test.go]: Sure

File: server/command.go
Comment (sibasankarnayak) [reply server/command.go]: removed by mistake while mocking the connection for localhost , will revert this change

**Comment by hanzei** [comment]
@sibasankarnayak Heads up that there are conflicts to resolve

**Comment by sibasankarnayak** [comment]
> 

@hanzei Done , resolved conflict

**Comment by hanzei** [comment]
@levb @mickmister Gentle reminder to review this PR

File: server/user.go
Comment (levb) [inline server/user.go:46]: why pointer values? what is the difference of a `nil` and a `false`?

**Review by levb — CHANGES_REQUESTED** [review]

**Comment by levb** [comment]
Also, I am not sure of the user value of having 3 separate settings, seems like a single setting is sufficient... @aaronrothschild

File: server/user.go
Comment (sibasankarnayak) [reply server/user.go:46]: @levb this was a suggestion from @mickmister https://github.com/mattermost/mattermost-plugin-jira/pull/705#discussion_r631444153 
it seems to keep a track where the user have explicitly turned off the notification.

Comment (mickmister) [reply server/user.go:46]: @levb We're using pointers for backwards compatibility, so we can fall back on the user's current settings if these are `nil`

**Comment by mickmister** [comment]
> Also, I am not sure of the user value of having 3 separate settings, seems like a single setting is sufficient

@levb Some of the notification types are noisier than others. Maybe a given user doesn't want watcher notifications because that's too many notifications. Note that the watcher notifications is introduced in another open PR, but the principle is the same. cc @aaronrothschild

**Comment by mickmister** [comment]
Some of the notification types are noisier than others. Maybe a given user doesn't want watcher 
notifications because that's too many notifications. Note that the watcher notifications is introduced in another open PR, but the principle is the same.

So if we factor in the new reporter notifications and watcher notifications, we have a total of 4 reasons why someone might get a notification:

- Mentioned
- Assignee
- Reporter
- Watcher

The question is if these should all be configured with one user setting `/jira settings notifications`, or if they should all be configured, i.e.

- `/jira settings notifications mention true`
- `/jira settings notifications assignee true`
- `/jira settings notifications reporter true`
- `/jira settings notifications watcher true`

**Comment by sibasankarnayak** [comment]
@mickmister made the changes as discussed over call , Please review and share the feedback.

**Comment by mickmister** [comment]
Since there is a lot of overlap with this PR and https://github.com/mattermost/mattermost-plugin-jira/pull/840, I'm holding off on reviewing the current changes in this PR.

@sibasankarnayak What would you like to do here? Are you wanting to merge one PR before the other, or close one of the PRs?

**Comment by sibasankarnayak** [comment]
> Since there is a lot of overlap with this PR and #840, I'm holding off on reviewing the current changes in this PR.
> 
> @sibasankarnayak What would you like to do here? Are you wanting to merge one PR before the other, or close one of the PRs?

I would like to be merge this PR and then can continue on the https://github.com/mattermost/mattermost-plugin-jira/pull/840  .
Can we do that ?

**Comment by sibasankarnayak** [comment]
> > Since there is a lot of overlap with this PR and #840, I'm holding off on reviewing the current changes in this PR.
> > @sibasankarnayak What would you like to do here? Are you wanting to merge one PR before the other, or close one of the PRs?
> 
> I would like to be merge this PR and then can continue on the #840 . Can we do that ?

@mickmister share you comments on this.

**Comment by mickmister** [comment]
> I would like to be merge this PR and then can continue on the #840 . Can we do that ?

@sibasankarnayak Sure. Many of my comments on #840 apply to this PR as well. Would you like me to create the same comments on this PR?

**Comment by mickmister** [comment]
@sibasankarnayak So you're implying we should close this PR in favor of the other PR? I'm fine with that if this is what you're suggesting

**Comment by sibasankarnayak** [comment]
> @sibasankarnayak So you're implying we should close this PR in favor of the other PR? I'm fine with that if this is what you're suggesting

@mickmister yes i mean that ,

will be closing this and continue working on the other PR to close reporter  and watching notification

**Pull request closed without merging**
